packages/dd-trace/src/profiling/exporters/agent.js

```diff
@@ -1,11 +1,58 @@
 'use strict'
 
+const http = require('http')
 const FormData = require('form-data')
 
+function wait (ms) {
+  return new Promise((resolve) => setTimeout(resolve, ms))
+}
+
+function * times (base, timeout) {
+  let attempt = 0
+  let n
+
+  // base * n is the current stride length
+  // base * (n - 1) is the sum prior steps
+  // therefore, base * n + base * (n - 1) is the final total
+  while (timeout > base * ((n = 2 ** attempt++) - 1) + base * n) {
+    yield base * n * Math.random()
+  }
+}
+
+async function backoff (base, overallTimeout, task) {
+  for (const timeout of times(base, overallTimeout)) {
+    const start = Date.now()
+    if (await task(timeout)) return
+    const remaining = timeout - (Date.now() - start)
+    if (remaining > 0) await wait(remaining)
+  }
+
+  throw new Error('Profiler agent export back-off failed')
+}
+
+function request (options, body) {
+  return new Promise((resolve, reject) => {
+    const request = http.request(options, resolve)
+    request.on('error', reject)
+    if (body) request.write(body)
+    request.end()
+  })
+}
+
+async function getBody (stream) {
+  const chunks = []
+  for await (const chunk of stream) {
```

Comment:
I'm getting a bit worried with the increasing number of promises since they tend to cause issues with `async_hooks` and by extension the tracer, and the GC is extremely lazy about them. There was a point in the life of `dd-trace` where there was a single place in the code using promises, and even then it ended up causing performance issues after being called enough times.

While promises are cleaner, in our case we have to prioritize performance and stability over cleaner code.

Reply:
I've eliminated more of the promises in the backoff mechanism, though I'm not convinced the performance would actually be noticeably different, especially given that this only runs once every minute. 🤷🏻